File: src/org/elixir_lang/ElixirLexer.java
package org.elixir_lang;

import com.intellij.lexer.FlexAdapter;
import com.intellij.lexer.LookAheadLexer;
import com.intellij.lexer.MergingLexerAdapter;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.TokenSet;
import org.elixir_lang.psi.ElixirTypes;

/**
 * Created by luke.imhoff on 8/3/14.
 */
public class ElixirLexer extends LookAheadLexer {
    public static final TokenSet FRAGMENTS = TokenSet.create(
            ElixirTypes.CHAR_LIST_FRAGMENT,
            ElixirTypes.SIGIL_FRAGMENT,
            ElixirTypes.STRING_FRAGMENT
    );

    public ElixirLexer() {
        super(
                new MergingLexerAdapter(
                        new ElixirFlexLexerAdapter(),
                        FRAGMENTS
                )
        );
    }
}

File: src/org/elixir_lang/ElixirLexer.kt
package org.elixir_lang

import com.intellij.lexer.LookAheadLexer
import com.intellij.lexer.MergingLexerAdapter
import com.intellij.openapi.project.Project
import com.intellij.psi.tree.TokenSet
import org.elixir_lang.psi.ElixirTypes

class ElixirLexer(private val elixirFlexLexerAdapter: ElixirFlexLexerAdapter) :
        LookAheadLexer(MergingLexerAdapter(elixirFlexLexerAdapter, FRAGMENTS)) {
    constructor(project: Project?): this(ElixirFlexLexerAdapter(project))
    constructor(): this(null)

    var level: Level?
      get() = elixirFlexLexerAdapter.level
      set(value) {
          elixirFlexLexerAdapter.level = value
      }

    fun stackSize(): Int = elixirFlexLexerAdapter.stackSize()

    companion object {
        val FRAGMENTS = TokenSet.create(
                ElixirTypes.ATOM_FRAGMENT,
                ElixirTypes.CHAR_LIST_FRAGMENT,
                ElixirTypes.REGEX_FRAGMENT,
                ElixirTypes.SIGIL_FRAGMENT,
                ElixirTypes.STRING_FRAGMENT,
                ElixirTypes.WORDS_FRAGMENT
        )
    }
}
